File: udp-py/udp/channel.py
# -*- coding: utf-8 -*-
#
#   UDP: User Datagram Protocol
#
#                                Written in 2020 by Moky <albert.moky@gmail.com>
#
# ==============================================================================
# MIT License
#
# Copyright (c) 2020 Albert Moky
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.
# ==============================================================================

import socket
import weakref
from typing import Optional, Tuple

from startrek import SocketAddress
from startrek import SocketReader, SocketWriter
from startrek import SocketHelper
from startrek import BaseChannel

from .aio import DatagramHelper


class ChannelChecker:

    @classmethod
    async def check_error(cls, error: OSError, sock: socket.socket) -> Optional[OSError]:
        """
            Check socket error

               (1) check E_AGAIN
                   the socket will raise 'Resource temporarily unavailable'
                   when received nothing in non-blocking mode,
                   or buffer overflow while sending too many bytes,
                   here we should ignore this exception.
               (2) check timeout
                   in blocking mode, the socket will wait until send/received data,
                   but if timeout was set, it will raise 'timeout' error on timeout,
                   here we should ignore this exception
        """
        # the socket will raise 'Resource temporarily unavailable'
        # when received nothing in non-blocking mode,
        # or buffer overflow while sending too many bytes,
        # here we should ignore this exception.
        if error.errno == socket.EAGAIN:  # error.strerror == 'Resource temporarily unavailable':
            helper = DatagramHelper()
            if not helper.is_blocking(sock=sock):
                # ignore it
                return None
        # in blocking mode, the socket wil wait until sent/received data,
        # but if timeout was set, it will raise 'timeout' error on timeout,
        # here we should ignore this exception.
        if isinstance(error, socket.timeout):
            if sock.gettimeout() is not None:  # or not self.blocking:
                # ignore it
                return None
        # print('[NET] socket error: %s' % error)
        return error

    @classmethod
    async def check_data(cls, data: Optional[bytes], sock: socket.socket) -> Optional[OSError]:
        """
            Check data received from socket

            (1) check timeout
                in blocking mode, the socket will wait until received something,
                but if timeout was set, it will return nothing too, it's normal;
                otherwise, we know the connection was lost.
        """
        # in blocking mode, the socket will wait until received something,
        # but if timeout was set, it will return None too, it's normal;
        # otherwise, we know the connection was lost.
        if data is None or len(data) == 0:
            if sock.gettimeout() is None:  # and self.blocking:
                # print('[NET] socket error: remote peer reset socket %s' % sock)
                return OSError('remote peer reset socket %s' % sock)


class Controller:
    """ Socket Channel Controller """

    def __init__(self, channel):
        super().__init__()
        self.__channel = weakref.ref(channel)
        self.__helper = DatagramHelper()

    @property
    def socket_helper(self) -> DatagramHelper:
        return self.__helper

    @property
    def channel(self):  # -> Optional[BaseChannel]:
        return self.__channel()

    @property
    def remote_address(self) -> Optional[SocketAddress]:
        channel = self.channel
        if isinstance(channel, BaseChannel):
            return channel.remote_address

    @property
    def local_address(self) -> Optional[SocketAddress]:
        channel = self.channel
        if isinstance(channel, BaseChannel):
            return channel.local_address

    @property
    def sock(self) -> Optional[socket.socket]:
        channel = self.channel
        if isinstance(channel, BaseChannel):
            return channel.sock


class PacketChannelReader(Controller, SocketReader):
    """ Datagram Packet Channel Reader """

    async def _socket_receive(self, sock: socket.socket, max_len: int) -> Optional[bytes]:
        helper = self.socket_helper
        if helper.is_closed(sock=sock):
            raise ConnectionError('socket closed')
        elif not helper.is_available(sock=sock):
            # TODO: check 'broken pipe'
            return None
        # elif helper.is_blocking(sock=sock):
        #     return sock.recv(max_len)
        else:
            return await helper.receive(sock=sock, max_len=max_len)  # raise OSError

    async def _socket_receive_from(self, sock: socket.socket, max_len: int
                                   ) -> Tuple[Optional[bytes], Optional[SocketAddress]]:
        helper = self.socket_helper
        if helper.is_closed(sock=sock):
            raise ConnectionError('socket closed')
        elif not helper.is_available(sock=sock):
            # TODO: check 'broken pipe'
            return None, None
        else:
            # return sock.recvfrom(max_len)
            return await helper.receive_from(sock=sock, max_len=max_len)  # raise OSError

    async def _try_read(self, max_len: int, sock: socket.socket) -> Optional[bytes]:
        try:
            # return sock.recv(max_len)
            return await self._socket_receive(sock=sock, max_len=max_len)  # raise OSError
        except OSError as error:
            error = await ChannelChecker.check_error(error=error, sock=sock)
            if error is None:
                # received nothing
                return None
            else:
                # connection lost?
                raise error

    # Override
    async def read(self, max_len: int) -> Optional[bytes]:
        sock = self.sock
        if sock is None:
            raise ConnectionError('channel not ready')
        # 1. try to read data
        data = await self._try_read(max_len=max_len, sock=sock)  # raise OSError
        # 2. check data
        error = await ChannelChecker.check_data(data=data, sock=sock)
        if error is not None:
            # connection lost!
            raise error
        # OK
        return data

    async def _try_receive(self, max_len: int, sock: socket.socket) -> Tuple[Optional[bytes], Optional[SocketAddress]]:
        try:
            # return sock.recvfrom(max_len)
            return await self._socket_receive_from(sock=sock, max_len=max_len)  # raise OSError
        except OSError as error:
            error = await ChannelChecker.check_error(error=error, sock=sock)
            if error is None:
                # received nothing
                return None, None
            else:
                # connection lost?
                raise error

    async def _receive_from(self, max_len: int) -> Tuple[Optional[bytes], Optional[SocketAddress]]:
        sock = self.sock
        if sock is None:
            raise ConnectionError('channel not ready')
        data, remote = await self._try_receive(max_len=max_len, sock=sock)  # raise OSError
        # check data
        error = await ChannelChecker.check_data(data=data, sock=sock)
        if error is None:
            # OK
            return data, remote
        else:
            # connection lost!
            raise error

    # Override
    async def receive(self, max_len: int) -> Tuple[Optional[bytes], Optional[SocketAddress]]:
        remote = self.remote_address
        if remote is None:
            # not connect (UDP)
            return await self._receive_from(max_len=max_len)  # raise OSError
        else:
            # connected (TCP/UDP)
            return await self.read(max_len=max_len), remote


class PacketChannelWriter(Controller, SocketWriter):
    """ Datagram Packet Channel Writer """

    async def _socket_send(self, sock: socket.socket, data: bytes) -> int:
        helper = self.socket_helper
        if helper.is_closed(sock=sock):
            raise ConnectionError('socket closed')
        # elif not helper.is_vacant(sock=sock):
        #     # TODO: check 'broken pipe'
        #     return -1
        # elif helper.is_blocking(sock=sock):
        #     return sock.send(data)
        else:
            return await helper.send(sock=sock, data=data)  # raise OSError

    async def _socket_send_to(self, sock: socket.socket, data: bytes, target: SocketAddress) -> int:
        helper = self.socket_helper
        if helper.is_closed(sock=sock):
            raise ConnectionError('socket closed')
        # elif not helper.is_vacant(sock=sock):
        #     # TODO: check 'broken pipe'
        #     return -1
        else:
            # return sock.sendto(data, remote)
            return await helper.send_to(sock=sock, data=data, target=target)  # raise OSError

    async def _try_write(self, data: bytes, sock: socket.socket) -> int:
        try:
            # return sock.send(data)
            return await self._socket_send(sock=sock, data=data)  # raise OSError
        except OSError as error:
            error = await ChannelChecker.check_error(error=error, sock=sock)
            if error is not None:
                # connection lost?
                raise error
            # buffer overflow!
            return 0

    # Override
    async def write(self, data: bytes) -> int:
        """ Return the number of bytes sent;
            this may be less than len(data) if the network is busy. """
        sock = self.sock
        if sock is None:
            raise ConnectionError('channel not ready')
        # sent = sock.sendall(data)
        sent = 0
        rest = len(data)
        # assert rest > 0, 'cannot send empty data'
        while True:  # while is_opened(sock=sock):
            cnt = await self._try_write(data=data, sock=sock)  # raise OSError
            # check send result
            if cnt <= 0:
                # buffer overflow?
                break
            # something sent, check remaining data
            sent += cnt
            rest -= cnt
            if rest <= 0:
                # done!
                break
            else:
                # remove sent part
                data = data[cnt:]
        # OK
        if sent > 0:
            return sent
        elif cnt < 0:
            assert cnt == -1, 'sent error: %d' % cnt
            return -1
        else:
            return 0

    async def _try_send(self, data: bytes, target: SocketAddress, sock: socket.socket) -> int:
        try:
            # return sock.sendto(data, target)
            return await self._socket_send_to(sock=sock, data=data, target=target)  # raise OSError
        except OSError as error:
            error = await ChannelChecker.check_error(error=error, sock=sock)
            if error is None:
                # buffer overflow!
                return -1
            else:
                # connection lost?
                raise error

    # Override
    async def send(self, data: bytes, target: SocketAddress) -> int:
        remote = self.remote_address
        if remote is None:
            # not connect (UDP)
            assert target is not None, 'target missed for unbound channel'
            return await self._try_send(data=data, target=target, sock=self.sock)  # raise OSError
        else:
            # connected (TCP/UDP)
            remote = self.remote_address
            assert target is None or target == remote, 'target error: %s, remote=%s' % (target, remote)
            # return sock.send(data)
            return await self._try_write(data=data, sock=self.sock)  # raise OSError


class PacketChannel(BaseChannel):
    """ Datagram Packet Channel """

    # Override
    def _create_reader(self) -> SocketReader:
        return PacketChannelReader(channel=self)

    # Override
    def _create_writer(self) -> SocketWriter:
        return PacketChannelWriter(channel=self)

    # Override
    def _create_helper(self) -> SocketHelper:
        return DatagramHelper()

File: startrek-js/StarTrek/src/socket/channel.js
;
// license: https://mit-license.org
//
//  Star Trek: Interstellar Transport
//
//                               Written in 2022 by Moky <albert.moky@gmail.com>
//
// =============================================================================
// The MIT License (MIT)
//
// Copyright (c) 2022 Albert Moky
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
// =============================================================================
//

//! require 'type/apm.js'
//! require 'controller.js'

/**
 *  WebSocket Wrapper
 *  ~~~~~~~~~~~~~~~~~
 *
 *  Socket = {
 *      configureBlocking: function(block) {},
 *
 *      isBlocking:  function() {return false},
 *      isOpen:      function() {return true},
 *      isConnected: function() {return true},
 *      isBound:     function() {return true},
 *      isAlive:     function() {return true},
 *
 *      getRemoteAddress: function() {return remote},
 *      getLocalAddress:  function() {return local},
 *
 *      bind:    function(local)  {return ws},
 *      connect: function(remote) {return ws},
 *      close:   function()       {ws.close()},
 *
 *      read:    function(maxLen)       {return new Uint8Array(maxLen)},
 *      write:   function(data)         {return data.length},
 *      receive: function(maxLen)       {return new Uint8Array(maxLen)},
 *      send:    function(data, remote) {return data.length}
 *  }
 */

(function (ns, sys) {
    'use strict';

    var Class = sys.type.Class;
    var AddressPairObject = ns.type.AddressPairObject;
    var Channel           = ns.net.Channel;
    var ChannelStateOrder = ns.net.ChannelStateOrder;
    var SocketHelper      = ns.net.SocketHelper;

    /**
     *  Base Channel
     *  ~~~~~~~~~~~~
     *
     * @param {SocketAddress} remote - remote address
     * @param {SocketAddress} local  - local address
     */
    var BaseChannel = function (remote, local) {
        AddressPairObject.call(this, remote, local);
        // SocketReader, SocketWriter
        this.__reader = this.createReader();
        this.__writer = this.createWriter();
        // inner socket
        this.__sock = null;  // WebSocket wrapper
        this.__closed = -1;  // 0: false, 1: true
    };
    Class(BaseChannel, AddressPairObject, [Channel], {

        // Override
        toString: function () {
            var clazz     = this.getClassName();
            var remote    = this.getRemoteAddress();
            var local     = this.getLocalAddress();
            var closed   = !this.isOpen();
            var bound     = this.isBound();
            var connected = this.isConnected();
            var sock      = this.getSocket();
            return '<' + clazz + ' remote="' + remote + '" local="' + local + '"' +
                ' closed=' + closed + ' bound=' + bound + ' connected=' + connected + '>\n\t' +
                sock + '\n</' + clazz + '>';
        }
    });

    // create socket reader & writer
    BaseChannel.prototype.createReader = function () {};
    BaseChannel.prototype.createWriter = function () {};
    // protected
    BaseChannel.prototype.getReader = function () {
        return this.__reader;
    };
    BaseChannel.prototype.getWriter = function () {
        return this.__writer;
    };

    //
    //  Socket
    //

    BaseChannel.prototype.getSocket = function () {
        return this.__sock;
    };

    /**
     *  Set inner socket for this channel
     *
     * @param {WebSocket} sock
     */
    BaseChannel.prototype.setSocket = function (sock) {
        // 1. replace with new socket
        var old = this.__sock;
        if (sock) {
            this.__sock = sock;
            this.__closed = 0;
        } else {
            this.__sock = null;
            this.__closed = 1;
        }
        // 2. close old socket
        if (old && old !== sock) {
            SocketHelper.socketDisconnect(old);
        }
    };

    //
    //  States
    //

    // Override
    BaseChannel.prototype.getState = function () {
        if (this.__closed < 0) {
            // initializing
            return ChannelStateOrder.INIT;
        }
        var sock = this.getSocket();
        if (!sock || SocketHelper.socketIsClosed(sock)) {
            // closed
            return ChannelStateOrder.CLOSED;
        } else if (SocketHelper.socketIsConnected(sock) || SocketHelper.socketIsBound(sock)) {
            // normal
            return ChannelStateOrder.ALIVE;
        } else {
            // opened
            return ChannelStateOrder.OPEN;
        }
    };

    // Override
    BaseChannel.prototype.isOpen = function () {
        if (this.__closed < 0) {
            // initializing
            return true;
        }
        var sock = this.getSocket();
        return sock && !SocketHelper.socketIsClosed(sock);
    };

    // Override
    BaseChannel.prototype.isBound = function () {
        var sock = this.getSocket();
        return sock && SocketHelper.socketIsBound(sock);
    };

    // Override
    BaseChannel.prototype.isConnected = function () {
        var sock = this.getSocket();
        return sock && SocketHelper.socketIsConnected(sock);
    };

    // Override
    BaseChannel.prototype.isAlive = function () {
        return this.isOpen() && (this.isConnected() || this.isBound());
    };

    // Override
    BaseChannel.prototype.isAvailable = function () {
        var sock = this.getSocket();
        if (!sock || SocketHelper.socketIsClosed(sock)) {
            return false;
        } else if (SocketHelper.socketIsConnected(sock) || SocketHelper.socketIsBound(sock)) {
            // alive, check reading buffer
            return this.checkAvailable(sock);
        } else {
            return false;
        }
    };

    // protected
    BaseChannel.prototype.checkAvailable = function (sock) {
        return SocketHelper.socketIsAvailable(sock);
    };

    // Override
    BaseChannel.prototype.isVacant = function () {
        var sock = this.getSocket();
        if (!sock || SocketHelper.socketIsClosed(sock)) {
            return false;
        } else if (SocketHelper.socketIsConnected(sock) || SocketHelper.socketIsBound(sock)) {
            // alive, check writing buffer
            return this.checkVacant(sock);
        } else {
            return false;
        }
    };

    // protected
    BaseChannel.prototype.checkVacant = function (sock) {
        return SocketHelper.socketIsVacant(sock);
    };

    // Override
    BaseChannel.prototype.isBlocking = function () {
        var sock = this.getSocket();
        return sock && SocketHelper.socketIsBlocking(sock);
    };

    // Override
    BaseChannel.prototype.configureBlocking = function (block) {
        var sock = this.getSocket();
        sock.configureBlocking(block);
        return sock;
    };

    // protected
    BaseChannel.prototype.doBind = function (sock, local) {
        return SocketHelper.socketBind(sock, local);
    };

    // protected
    BaseChannel.prototype.doConnect = function (sock, remote) {
        return SocketHelper.socketConnect(sock, remote);
    };

    // protected
    BaseChannel.prototype.doDisconnect = function (sock) {
        return SocketHelper.socketDisconnect(sock);
    };

    // Override
    BaseChannel.prototype.bind = function (local) {
        var sock = this.getSocket();
        if (sock) {
            this.doBind(sock, local);
        }
        this.localAddress = local;
        return sock;
    };

    // Override
    BaseChannel.prototype.connect = function (remote) {
        var sock = this.getSocket();
        if (sock) {
            this.doConnect(sock, remote);
        }
        this.remoteAddress = remote;
        return sock;
    };

    // Override
    BaseChannel.prototype.disconnect = function () {
        var sock = this.getSocket();
        if (sock) {
            this.doDisconnect(sock);
        }
        return sock;
    };

    // Override
    BaseChannel.prototype.close = function () {
        this.setSocket(null);
    };

    //
    //  Reading, Writing
    //

    // Override
    BaseChannel.prototype.read = function (maxLen) {
        try {
            return this.getReader().read(maxLen);
        } catch (e) {
            this.close();
            throw e;
        }
    };

    // Override
    BaseChannel.prototype.write = function (src) {
        try {
            return this.getWriter().write(src);
        } catch (e) {
            this.close();
            throw e;
        }
    };

    // Override
    BaseChannel.prototype.receive = function (maxLen) {
        try {
            return this.getReader().receive(maxLen);
        } catch (e) {
            this.close();
            throw e;
        }
    };

    // Override
    BaseChannel.prototype.send = function (src, target) {
        try {
            return this.getWriter().send(src, target);
        } catch (e) {
            this.close();
            throw e;
        }
    };

    //-------- namespace --------
    ns.socket.BaseChannel = BaseChannel;

})(StarTrek, MONKEY);
